error messages: condition chaining

becker33 opened the pull request:
fixes #40290

Supersedes #34500 

This is work to create chains of causation for error messages.

The current implementation is only completed for some of the many errors presented by the concretizer. The rest will need to be filled out over time, but this demonstrates the capability. 

The basic idea is to associate conditions in the solver with one another in causal relationships, and to associate errors with the proximate causes of their facts in the condition graph. Then we can construct causal trees to explain errors, which will hopefully present users with useful information to avoid the error or report issues.

Examples:

```
$ spack solve hdf5 ^cmake@3.0.1
==> Error: concretization failed for the following reasons:

   1. Cannot satisfy 'cmake@3.0.1'
   2. Cannot satisfy 'cmake@3.0.1'
        required because hdf5 ^cmake@3.0.1 requested from CLI 
   3. Cannot satisfy 'cmake@3.18:' and 'cmake@3.0.1
        required because hdf5 ^cmake@3.0.1 requested from CLI 
        required because hdf5 depends on cmake@3.18: when @1.13: 
          required because hdf5 ^cmake@3.0.1 requested from CLI 
   4. Cannot satisfy 'cmake@3.12:' and 'cmake@3.0.1
        required because hdf5 depends on cmake@3.12: 
          required because hdf5 ^cmake@3.0.1 requested from CLI 
        required because hdf5 ^cmake@3.0.1 requested from CLI
```

```
$ spack spec cmake ^curl~ldap   # <-- with curl configured non-buildable and an external with `+ldap`
==> Error: concretization failed for the following reasons:

   1. Attempted to use external for 'curl' which does not satisfy any configured external spec
   2. Attempted to build package curl which is not buildable and does not have a satisfying external
        attr('variant_value', 'curl', 'ldap', 'True') is an external constraint for curl which was not satisfied
   3. Attempted to build package curl which is not buildable and does not have a satisfying external
        attr('variant_value', 'curl', 'gssapi', 'True') is an external constraint for curl which was not satisfied
   4. Attempted to build package curl which is not buildable and does not have a satisfying external
        'curl+ldap' is an external constraint for curl which was not satisfied
        'curl~ldap' required
        required because cmake ^curl~ldap requested from CLI 
```

```
$ spack solve yambo+mpi ^hdf5~mpi
==> Error: concretization failed for the following reasons:

   1. 'hdf5' required multiple values for single-valued variant 'mpi'
   2. 'hdf5' required multiple values for single-valued variant 'mpi'
    Requested '~mpi' and '+mpi'
        required because yambo depends on hdf5+mpi when +mpi 
          required because yambo+mpi ^hdf5~mpi requested from CLI 
        required because yambo+mpi ^hdf5~mpi requested from CLI 
   3. 'hdf5' required multiple values for single-valued variant 'mpi'
    Requested '~mpi' and '+mpi'
        required because netcdf-c depends on hdf5+mpi when +mpi 
          required because netcdf-fortran depends on netcdf-c 
            required because yambo depends on netcdf-fortran 
              required because yambo+mpi ^hdf5~mpi requested from CLI 
          required because netcdf-fortran depends on netcdf-c@4.7.4: when @4.5.3: 
            required because yambo depends on netcdf-fortran 
              required because yambo+mpi ^hdf5~mpi requested from CLI 
          required because yambo depends on netcdf-c 
            required because yambo+mpi ^hdf5~mpi requested from CLI 
          required because yambo depends on netcdf-c+mpi when +mpi 
            required because yambo+mpi ^hdf5~mpi requested from CLI 
        required because yambo+mpi ^hdf5~mpi requested from CLI 
```

In addition to fleshing out the causes of other errors, I would like to find a way to associate different components of the error messages with different causes. In this example it's pretty easy to infer which part is which, but I'm not confident that will always be the case. But that's for future work.

~~Due to performance degradation concerns, the condition chaining work is implemented as a separate file that can be loaded only in debug mode. While the PR is WIP the more verbose error message are configured to be default because it's easier to work on that way.~~

See the previous PR #34500 for discussion of how the condition chains are incomplete. Given the improvements here, I think it's worth trying to get this in first before worrying too much about the cases that we aren't yet able to pick up on, but I also want to be sure to acknowledge that those exist.

becker33 committed:
add internal_error clauses to many integrity constraints

becker33 committed:
create causal chains; test on one error message

becker33 committed:
turn literals into conditions so we can track causation back to them

becker33 committed:
explanatory message for requirement conditions

becker33 committed:
asp.py: allow condition creator to take explicit requirements/impositions instead of specs

becker33 committed:
solver: turn dependency conditions into normal conditions

becker33 committed:
solver: turn external conditions into normal conditions

becker33 committed:
error_messages: 2-cause error message for version satisfaction

becker33 committed:
solver: make dependency condition messages more verbose

becker33 committed:
deprioritize 1-cause error message for versions in favor of 2-cause message

becker33 committed:
special condition_cause derivation for dependencies

becker33 committed:
error message with causes for multiple values on single-valued variant

becker33 committed:
loosen requirements on condition chaining inferences matching attr packages

becker33 committed:
cause-based error messages for externals

becker33 committed:
show additional causation for variants preventing external use

becker33 committed:
style

becker33 committed:
remove heuristics for literals

alalazo reviewed (changes requested):
I don't think the performance of this PR is sustainable in the current state :slightly_frowning_face:

 [The pipelines](https://gitlab.spack.io/spack/spack/-/pipelines/506036) couldn't finish concretization in an hour.

I'm currently trying to time a small benchmark, but I don't know how long it will take:
![Screenshot from 2023-09-28 10-18-20](https://github.com/spack/spack/assets/4199709/71c095dc-ad84-4a60-8716-cb0449a9946a)

`aluminum` is taking 12s on `develop` and `ascent` is taking 31s as a reference.

If the differences stays around 20x slow-down, we need at the very least to have an ad-hoc flag for users to turn this on (if we turn it on during debug, it might seem that the debug run is hanging).

haampie commented:
Also memory usage seems to be 2x (in the `hdf5 ^cmake@3.0.1` example)

develop

```
Max PSS: 706’966’528 bytes
Walltime: 12’987ms - user-space: 12’361ms - kernel-space: 356ms
706┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
   ┃                                                                ███████████████      ┃
   ┃                                                              ██████████████████     ┃
   ┃                                                              ██████████████████████ ┃
   ┃                                                              ██████████████████████ ┃
   ┃                                                            ████████████████████████ ┃
   ┃                                                     ███████████████████████████████ ┃
   ┃                                                   █████████████████████████████████ ┃
   ┫                                                 ████████████████████████████████████┃
   ┃                                               ██████████████████████████████████████┃
   ┃                                             ████████████████████████████████████████┃
   ┃                                           ██████████████████████████████████████████┃
   ┃                                     ████████████████████████████████████████████████┃
   ┃          ███████████████████████████████████████████████████████████████████████████┃
   ┃  ███████████████████████████████████████████████████████████████████████████████████┃
   ┃█████████████████████████████████████████████████████████████████████████████████████┃
0MB┗━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛
   0s                                                                                   12
```

pr

```
Max PSS: 1’491’908’608 bytes
Walltime: 20’838ms - user-space: 19’401ms - kernel-space: 681ms
  1┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
   ┃                                                                                     ┃
   ┃                                                  ████████████████████████████████   ┃
   ┃                                                 ███████████████████████████████████ ┃
   ┃                                                ████████████████████████████████████ ┃
   ┃                                                ████████████████████████████████████ ┃
   ┃                                                ████████████████████████████████████ ┃
   ┃                                               █████████████████████████████████████ ┃
   ┫                                    ████████████████████████████████████████████████ ┃
   ┃                                █████████████████████████████████████████████████████┃
   ┃                              ███████████████████████████████████████████████████████┃
   ┃                           ██████████████████████████████████████████████████████████┃
   ┃                        █████████████████████████████████████████████████████████████┃
   ┃                     ████████████████████████████████████████████████████████████████┃
   ┃       ██████████████████████████████████████████████████████████████████████████████┃
   ┃█████████████████████████████████████████████████████████████████████████████████████┃
0GB┗━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛
   0s                                                                                   20
```

alalazo commented:
Tested on:

* **Spack:** 0.21.0.dev0 (1d6394bec96622ab8689424b8a9084d8d9e38987)
* **Python:** 3.11.4
* **Platform:** linux-ubuntu20.04-icelake
* **Concretizer:** clingo

PR: 1d6394bec96622ab8689424b8a9084d8d9e38987
develop: ebc0e9be1971c5670b6a9f9d6ffd0b4b1c2354bb

[radiuss-error-message.csv](https://github.com/spack/spack/files/12747985/radiuss-error-message.csv)
[radiuss-develop.csv](https://github.com/spack/spack/files/12747986/radiuss-develop.csv)
[radiuss.txt](https://github.com/spack/spack/files/12747987/radiuss.txt)

This is a plot of the results, in log scale:

![radiuss](https://github.com/spack/spack/assets/4199709/6abb5016-df8e-43f4-8fd4-19c46c3701df)

alalazo commented:
Here's my high-level thoughts on the architecture of this. Currently we are throwing additional rules in the solve to get good error messages. These rules compute *all possible causation paths from all possible solutions*. This is likely not what we want, and it leads to the performance issue in https://github.com/spack/spack/pull/40173#issuecomment-1738821710, which is quite severe.

I think the objective of this should just be to explain an error when it occurs. Hence the idea:
- Can we turn this into a **separate** logic problem whose input is a solution that contains an `error` fact?

If we could do so we would greatly reduce the search space, cause it wouldn't be anymore "all possible solutions" from the solve - but just the "optimal" solution with an error.

adamjstewart commented:
In my limited conda experience, when conda is able to concretize, it does so quickly. But when something is wrong, conda takes 30+ minutes to generate an error. I'm guessing this is similar to what @alalazo is suggesting. Concretization speed shouldn't be affected unless there's an actual issue.

becker33 committed:
create secondary solve for error messages only when error occurs

becker33 commented:
The performance problem should be mostly addressed -- I haven't done extensive testing, but from what I'm seeing in quick testing I'm seeing ~10% performance penalties on successful concretizations and ~20% performance penalties on unsuccessful concretizations. This improvement comes from following @alalazo 's suggestion to decouple the error causation calculation from the rest of the solve.

@alalazo can you redo the performance analysis or either give me or remind me where I can find the script you use?

becker33 committed:
cleanup previous commit

alalazo commented:
> can you redo the performance analysis or either give me or remind me where I can find the script you use?

I'll re-post timings here. To do produce them, I have a script to do the plots and a [spack extension](https://github.com/alalazo/spack-benchmark) to produce the CSV files

becker33 committed:
separate display for error messages from display for model

becker33 committed:
show causation for conflicts

becker33 committed:
error message with causes for virtual providers

becker33 committed:
remove vestigial variable

becker33 committed:
simplify unnecessarily complex method to add model facts to error solve

alalazo commented:
Measured again the same benchmark:

* **Spack:** 0.21.0.dev0 (b2d5084d948a418480d86c7ad3592160d0d3d342)
* **Python:** 3.11.4
* **Platform:** linux-ubuntu20.04-icelake
* **Concretizer:** clingo

develop: 08da9a854a53542e1f71d30ed8d14d6b7d18119f
PR: b2d5084d948a418480d86c7ad3592160d0d3d342

[radiuss-error-message-baseline.csv](https://github.com/spack/spack/files/12843592/radiuss-error-message-baseline.csv)
[radiuss-develop.csv](https://github.com/spack/spack/files/12843593/radiuss-develop.csv)
[radiuss.txt](https://github.com/spack/spack/files/12843594/radiuss.txt)

Here's is the plot:

![radiuss](https://github.com/spack/spack/assets/4199709/714f09c7-34f9-4e65-9f1e-3be4db57b712)

In terms of performance, this seems fine to me now. Given this, we can probably enable error messages by default :slightly_smiling_face: (not only in debug mode)

haampie commented:
Also peak memory stays roughly the same:

before
```
Max PSS: 1’556’853’760 bytes
Walltime: 32’289ms - user-space: 30’432ms - kernel-space: 663ms
  1┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
   ┃                                                             ████████████████████    ┃
   ┃                                                             ███████████████████████ ┃
   ┃                                                            ████████████████████████ ┃
   ┃                                                            ████████████████████████ ┃
   ┃                                                            ████████████████████████ ┃
   ┃                                                          ███████████████████████████┃
   ┃                                                 ████████████████████████████████████┃
   ┫                                                █████████████████████████████████████┃
   ┃                                              ███████████████████████████████████████┃
   ┃                                           ██████████████████████████████████████████┃
   ┃                                        █████████████████████████████████████████████┃
   ┃                                      ███████████████████████████████████████████████┃
   ┃                                    █████████████████████████████████████████████████┃
   ┃     ████████████████████████████████████████████████████████████████████████████████┃
   ┃█████████████████████████████████████████████████████████████████████████████████████┃
0GB┗━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛
   0s                                                                                   32
```

after
```
Max PSS: 1’591’031’808 bytes
Walltime: 35’002ms - user-space: 32’551ms - kernel-space: 834ms
  1┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
   ┃                                                                            █████    ┃
   ┃                                                         █████████████████████████   ┃
   ┃                                                        ████████████████████████████ ┃
   ┃                                                        ████████████████████████████ ┃
   ┃                                                       █████████████████████████████ ┃
   ┃                                                  ███████████████████████████████████┃
   ┃                                              ███████████████████████████████████████┃
   ┫                                            █████████████████████████████████████████┃
   ┃                                          ███████████████████████████████████████████┃
   ┃                                        █████████████████████████████████████████████┃
   ┃                                      ███████████████████████████████████████████████┃
   ┃                                    █████████████████████████████████████████████████┃
   ┃                                  ███████████████████████████████████████████████████┃
   ┃     ████████████████████████████████████████████████████████████████████████████████┃
   ┃█████████████████████████████████████████████████████████████████████████████████████┃
0GB┗━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛
   0s                                                                                   35
```

alalazo commented on lib/spack/spack/solver/asp.py:
```diff
@@ -1272,6 +1324,13 @@ def pkg_rules(self, pkg, tests):
         self.trigger_rules()
         self.effect_rules()
 
+        # dependencies have special triggers/effects, have to be handled separately
+        self.package_dependencies_rules(pkg)
+
+        # trigger and effect tables
+        self.trigger_rules()
+        self.effect_rules()
```
I'm still going through the PR, but I tried to revert this:
```diff
diff --git a/lib/spack/spack/solver/asp.py b/lib/spack/spack/solver/asp.py
index f4775a7dfa..f40b77d4f1 100644
--- a/lib/spack/spack/solver/asp.py
+++ b/lib/spack/spack/solver/asp.py
@@ -1312,6 +1312,9 @@ def pkg_rules(self, pkg, tests):
         # virtuals
         self.package_provider_rules(pkg)
 
+        # dependencies
+        self.package_dependencies_rules(pkg)
+
         # virtual preferences
         self.virtual_preferences(
             pkg.name,
@@ -1324,13 +1327,6 @@ def pkg_rules(self, pkg, tests):
         self.trigger_rules()
         self.effect_rules()
 
-        # dependencies have special triggers/effects, have to be handled separately
-        self.package_dependencies_rules(pkg)
-
-        # trigger and effect tables
-        self.trigger_rules()
-        self.effect_rules()
-
     def trigger_rules(self):
         """Flushes all the trigger rules collected so far, and clears the cache."""
         self.gen.h2("Trigger conditions")
```
and test passes. Also, from a quick inspection of the facts they seem to be the same (just rearranged), so not sure at the moment what is special about dependencies :thinking:

becker33 replied:
This was left-over from previous implementation, I'm reverting it. The caching already accounts for what's different about them.

alalazo commented on lib/spack/spack/solver/asp.py:
```diff
@@ -1405,42 +1464,49 @@ def condition(self, required_spec, imposed_spec=None, name=None, msg=None, node=
+        if isinstance(named_cond, spack.spec.Spec):
+            named_cond.name = named_cond.name or name
+            name = named_cond.name
+        assert name, "must provide name for anonymous conditions!"
```
Hmm, `required_spec` is supposed to be a `Spec` according to docs. Which other types are we passing that support a `.copy()` method? It would be good if we start adding type hints here, but at least we need to update docstrings.

alalazo commented on lib/spack/spack/solver/asp.py:
```diff
@@ -1486,21 +1552,24 @@ def package_dependencies_rules(self, pkg):
-                self.gen.fact(
-                    fn.pkg_fact(pkg.name, fn.dependency_condition(condition_id, dep.spec.name))
-                )
+                required_spec = cond.copy()
+                required_spec.name = cond.name or pkg.name
+                required = self.spec_clauses(required_spec, body=True)
+                required += [fn.attr("track_dependencies", pkg.name)]
+                imposed = self.spec_clauses(dep.spec)
+                imposed += [
+                    fn.attr("dependency_holds", pkg.name, dep.spec.name, dt.flag_to_string(t))
+                    for t in dt.ALL_FLAGS
+                    if t & depflag
+                ]
```
Hmm, I see here that the other type is `AspFunction`. I think this is mixing two different levels of abstraction and should be avoided (a spec is a higher level of talking about a requirement, than a series of facts). 

What is preventing to leave `self.condition` with the same signature, and avoid it being a function that branches on the type of arguments it gets?

becker33 replied:
the issue is that in some cases we have requirements or imposed constraints that are not directly part of the spec. So the idea with the change to the interface is that you can either give it a spec or manually construct the requirements or imposed_constraints if you need to do something more complicated with them.

I will see if I can make that clearer, and make sure it's still necessary

alalazo replied:
> So the idea with the change to the interface is that you can either give it a spec or manually construct the requirements or imposed_constraints if you need to do something more complicated with them.

I think it's clearer if we can inject a transformation function into `self.condition`, which takes a spec + a list of facts as input and gives a list of facts as output.

alalazo commented on lib/spack/spack/solver/asp.py:
```diff
@@ -1486,21 +1552,24 @@ def package_dependencies_rules(self, pkg):
-                    if t & depflag:
-                        # there is a declared dependency of type t
-                        self.gen.fact(fn.dependency_type(condition_id, dt.flag_to_string(t)))
+                _ = self.condition(required, imposed, pkg.name, msg)
```
This is a sign that the function now does two unrelated things: we call it (with different arguments) and we don't need the result anymore. Same happens below.

becker33 replied:
You only need the result if you're adding something else about the condition, many of the conditions we don't use the return value already.

alalazo commented on lib/spack/spack/solver/asp.py:
```diff
@@ -1659,8 +1732,8 @@ def external_packages(self):
-                condition_id = self.condition(spec, msg=msg)
-                self.gen.fact(fn.pkg_fact(pkg_name, fn.possible_external(condition_id, local_idx)))
+                external_imposition = [fn.attr("external_conditions_hold", spec.name, local_idx)]
+                _ = self.condition(spec, external_imposition, msg=msg)
```
:point_up:

alalazo commented on lib/spack/spack/solver/concretize.lp:
```diff
@@ -30,27 +30,25 @@
-:- provider(PackageNode, _), not attr("node", PackageNode).
-
-:- attr("root", node(ID, PackageNode)), ID > min_dupe_id.
+:- attr("root", node(ID, PackageNode)), ID > min_dupe_id, internal_error("root in non-minimum unification set").
```
This is about the ID of the node, not the unification set, so I think the message should be something like:
```suggestion
:- attr("root", node(ID, PackageNode)), ID > min_dupe_id, internal_error("root with a non-minimal duplicate ID").
```

alalazo commented on lib/spack/spack/solver/concretize.lp:
```diff
@@ -30,27 +30,25 @@
 % Rules on "unification sets", i.e. on sets of nodes allowing a single configuration of any given package
 unify(SetID, PackageName) :- unification_set(SetID, node(_, PackageName)).
-:- 2 { unification_set(SetID, node(_, PackageName)) }, unify(SetID, PackageName).
+:- 2 { unification_set(SetID, node(_, PackageName)) }, unify(SetID, PackageName), internal_error("unified package in multiple unification sets").
```
This is the opposite (multiple configuration of a package in the same unification set):
```suggestion
:- 2 { unification_set(SetID, node(_, PackageName)) }, unify(SetID, PackageName), internal_error("a package appears with two or more configuration in the same unification set").
```
In my opinion we can also avoid adding this `internal_error`.

alalazo commented on lib/spack/spack/solver/concretize.lp:
```diff
@@ -114,25 +114,47 @@ multiple_nodes_attribute("node_flag_source").
+  trigger_condition_holds(TriggerID, Node),
+  literal(TriggerID).
+
+mentioned_in_literal(Root, Mentioned) :-
+  literal(TriggerID),
+  trigger_and_effect(Root, TriggerID, EffectID),
+  imposed_constraint(EffectID, _, Mentioned).
+mentioned_in_literal(Root, Mentioned) :-
+  literal(TriggerID),
+  trigger_and_effect(Root, TriggerID, EffectID),
+  imposed_constraint(EffectID, _, Mentioned, _).
+mentioned_in_literal(Root, Mentioned) :-
+  literal(TriggerID),
+  trigger_and_effect(Root, TriggerID, EffectID),
+  imposed_constraint(EffectID, _, Mentioned, _, _).
+mentioned_in_literal(Root, Mentioned) :-
+  literal(TriggerID),
+  trigger_and_effect(Root, TriggerID, EffectID),
+  imposed_constraint(EffectID, _, Mentioned, _, _, _).
+mentioned_in_literal(Root, Mentioned) :-
+  literal(TriggerID),
+  trigger_and_effect(Root, TriggerID, EffectID),
+  imposed_constraint(EffectID, "depends_on", _, _, Mentioned, _).
```
These are mostly different by "arity". Wondering if we can compact these rules.

becker33 replied:
I'm pretty sure we cannot, unless we want to make all the arguements to `attr` be tuples instead of individual elements.

alalazo replied:
I believe this can be computed much more compactly in `asp.py`, when we emit literals. I'll give it a try and push here if I am successful.

alalazo replied:
@becker33 I just pushed https://github.com/spack/spack/pull/40173/commits/a2b6ef84cbbfbb344e2d1cc3fe00f58e1f6e3979 which computes this fact directly from `asp.py`.

alalazo commented on lib/spack/spack/solver/concretize.lp:
```diff
@@ -427,24 +451,11 @@ depends_on(PackageNode, DependencyNode) :- attr("depends_on", PackageNode, Depen
 % If a dependency holds on a package node, there must be one and only one dependency node satisfying it
 1 { attr("depends_on", PackageNode, node(0..Y-1, Dependency), Type) : max_dupes(Dependency, Y) } 1
-  :- dependency_holds(PackageNode, Dependency, Type),
+  :- attr("dependency_holds", PackageNode, Dependency, Type),
```
Why are we turning this into an attribute? Is it necessary?

becker33 replied:
We're turning it into an attribute so that it can be imposed by a dependency condition (so that we can connect dependency conditions the same way we connect other types of conditions

alalazo commented on lib/spack/spack/solver/heuristic.lp:
```diff
@@ -11,10 +11,11 @@
-#heuristic attr("root", node(0, Package)) : literal(_, "root", Package). [45, true]
-#heuristic attr("node", node(0, Package)) : literal(_, "root", Package). [45, true]
-#heuristic attr("node", node(0, Package)) : literal(_, "node", Package). [45, true]
+% TODO: see what I can salvage from these
+%#heuristic attr("hash", node(0, Package), Hash) : literal(_, "root", Package). [45, init]
+%#heuristic attr("root", node(0, Package)) : literal(_, "root", Package). [45, true]
+%#heuristic attr("node", node(0, Package)) : literal(_, "root", Package). [45, true]
+%#heuristic attr("node", node(0, Package)) : literal(_, "node", Package). [45, true]
```
:point_up:

alalazo commented on lib/spack/spack/solver/error_messages.lp:
```diff
@@ -0,0 +1,216 @@
+
+% Create a causal tree between trigger conditions by locating the effect conditions
+% that are triggers for another condition. Condition2 is caused by Condition1
+condition_cause(Condition2, Condition1, ID) :-
```
I think we should rewrite as:
```suggestion
condition_cause(EffectID, CauseID, PackageID) :-
```
I'm also wondering if we should pass the full `node(PackageID, Package)` for clarity (instead of only `PackageID`).

alalazo replied:
So, if I get this correctly, `Condition2` is caused by `Condition1` through `node(ID, Package)`. If that's true I think we might need:
1. To store the full node information
2. To have a better connection between the two condition than just the name of the node (this might come later)

For point 2 I think we might end up with wrong reconstructions if we have multiple nodes from the same package.

alalazo replied:
Another question (not necessarily a request): aren't we losing connections where the name of the attribute varies between the `condition_requirement` and the `imposed_constraint`? For instance, thinking of `variant_value` vs. `variant_set`? Should we have a map of those names?

becker33 replied:
Yes, that's probably something to address in the future. Off the top of my head I'm not thinking of another attribute name for which that would occur, but I expect we will find more.

alalazo replied:
It's probably these attributes:
```
node_platform = fn.attr("node_platform_set")
node_os = fn.attr("node_os_set")
node_target = fn.attr("node_target_set")
variant_value = fn.attr("variant_set")
node_compiler = fn.attr("node_compiler_set")
node_compiler_version = fn.attr("node_compiler_version_set")
node_flag = fn.attr("node_flag_set")
```

alalazo commented on lib/spack/spack/solver/error_messages.lp:
```diff
@@ -0,0 +1,216 @@
+  attr(Name, node(ID, Package), A1, A2, A3),
+  condition_holds(Condition1, node(ID1, Package1)),
+  pkg_fact(Package1, condition_effect(Condition1, Effect)),
+  imposed_constraint(Effect, Name, Package, A1, A2, A3).
```
These rules are also differing by arity. I think also the premise is wrong: why should the arity of the trigger condition go along with the arity of the imposed condition? Aren't we losing combinations?

alalazo replied:
We need to look into ways of compacting this, otherwise it would be like writing C++ templates without variadic args :laughing:

becker33 replied:
I think our in-person conversation cleared up that this is necessary, because only a single attribute (which implies single arity) can connect two conditions (whereas condition requirements and impositions need to be cross-arity, and are designed to be).

alalazo reviewed (changes requested):
Left a few comments. I still need to go through the rules for errors in `error_messages.lp`. I'll do that likely on Monday, and will also try to think about ways to simplify:
- The `condition` method API
- The rules that are replicated with different arity

becker33 committed:
clean up condition interface

becker33 committed:
clean up new internal errors in concretize.lp

becker33 committed:
revert unnecessary secondary trigger and effect tables from c62cb587e2

becker33 committed:
remove no-longer-valid heuristic for min_dupe_id literals as it's already forced

becker33 committed:
fixup 87efcf4ada

alalazo commented on lib/spack/spack/solver/asp.py:
```diff
@@ -1404,43 +1464,51 @@ def condition(self, required_spec, imposed_spec=None, name=None, msg=None, node=
         if named_cond_key not in cache:
             trigger_id = next(self._trigger_id_counter)
-            requirements = self.spec_clauses(named_cond, body=True, required_from=name)
+            if isinstance(named_cond, spack.spec.Spec):
+                requirements = self.spec_clauses(named_cond, body=True, required_from=name)
+            else:
+                requirements = named_cond
```
I still think this is confusing, and mixing abstraction levels. We should find a way to split it in a neater way, without if/else here and with the assumption that client code is responsible to call `spec_clauses`.

alalazo replied:
One idea is to change:
```python
def condition(self, required_spec, imposed_spec=None, name=None, msg=None, node=False):
```
into:
```python
def condition(self, required_spec, imposed_spec=None, name=None, msg=None, transform_requirements=None, transform_imposed=None):
```
where we _inject_ the transformation functions into condition, if we need to add anything. I'll try that out and report in a comment.

alalazo commented on lib/spack/spack/solver/asp.py:
```diff
@@ -1404,43 +1464,51 @@ def condition(self, required_spec, imposed_spec=None, name=None, msg=None, node=
         Returns:
             int: id of the condition created by this function
         """
+        # Defensive copy of string or spec
```
(minor) Wrong comment (it's not a string). I'd just leave it out:
```suggestion
```

alalazo reviewed (changes requested):
I still need to re-review the logic program, but leaving a first few comments. 

There's currently an import issue:
```console
$ spack spec yambo+mpi ^hdf5~mpi
==> Error: name 'pathlib' is not defined
```
See https://github.com/spack/spack/actions/runs/6713744850/job/18245826720?pr=40173

When running locally, a unit test fails:
```
lib/spack/spack/test/concretize_requirements.py::test_require_cflags
```

Also, this new functionality completely lacks unit tests - so we are more subject to undetected regressions in the future.

alalazo commented:
@becker33 I suggest the following diff to improve the implementation of `self.condition` in the solver:

<details>

<summary>Inject transformation functions</summary>

```diff
diff --git a/lib/spack/spack/solver/asp.py b/lib/spack/spack/solver/asp.py
index 2f6efc6469..8c96fd4586 100644
--- a/lib/spack/spack/solver/asp.py
+++ b/lib/spack/spack/solver/asp.py
@@ -8,11 +8,12 @@
 import enum
 import itertools
 import os
+import pathlib
 import pprint
 import re
 import types
 import warnings
-from typing import Dict, List, NamedTuple, Optional, Sequence, Tuple, Union
+from typing import Callable, Dict, List, NamedTuple, Optional, Sequence, Tuple, Union
 
 import archspec.cpu
 
@@ -337,6 +338,13 @@ def __getattr__(self, name):
 
 fn = AspFunctionBuilder()
 
+TransformFunction = Callable[[spack.spec.Spec, List[AspFunction]], List[AspFunction]]
+
+
+def remove_node(spec: spack.spec.Spec, facts: List[AspFunction]) -> List[AspFunction]:
+    """Transformation that removes all "node" and "virtual_node" from the input list of facts."""
+    return list(filter(lambda x: x.args[0] not in ("node", "virtual_node"), facts))
+
 
 def _create_counter(specs, tests):
     strategy = spack.config.CONFIG.get("concretizer:duplicates:strategy", "none")
@@ -1168,7 +1176,7 @@ def conflict_rules(self, pkg):
         default_msg = "{0}: '{1}' conflicts with '{2}'"
         no_constraint_msg = "{0}: conflicts with '{1}'"
         for trigger, constraints in pkg.conflicts.items():
-            trigger_msg = "conflict is triggered when %s" % str(trigger)
+            trigger_msg = f"conflict is triggered when {str(trigger)}"
             trigger_spec = spack.spec.Spec(trigger)
             trigger_id = self.condition(
                 trigger_spec, name=trigger_spec.name or pkg.name, msg=trigger_msg
@@ -1181,9 +1189,8 @@ def conflict_rules(self, pkg):
                     else:
                         conflict_msg = default_msg.format(pkg.name, trigger, constraint)
 
-                constraint_msg = "conflict applies to spec %s" % str(
-                    spack.spec.Spec(pkg.name) if constraint == spack.spec.Spec() else constraint
-                )
+                spec_for_msg = spack.spec.Spec(pkg.name) if constraint == spack.spec.Spec() else constraint
+                constraint_msg = f"conflict applies to spec {str(spec_for_msg)}"
                 constraint_id = self.condition(constraint, name=pkg.name, msg=constraint_msg)
                 self.gen.fact(
                     fn.pkg_fact(pkg.name, fn.conflict(trigger_id, constraint_id, conflict_msg))
@@ -1365,7 +1372,7 @@ def trigger_rules(self):
         self.gen.h2("Trigger conditions")
         for name in self._trigger_cache:
             cache = self._trigger_cache[name]
-            for spec_str, (trigger_id, requirements) in cache.items():
+            for (spec_str, _), (trigger_id, requirements) in cache.items():
                 self.gen.fact(fn.pkg_fact(name, fn.trigger_id(trigger_id)))
                 self.gen.fact(fn.pkg_fact(name, fn.trigger_msg(spec_str)))
                 for predicate in requirements:
@@ -1378,7 +1385,7 @@ def effect_rules(self):
         self.gen.h2("Imposed requirements")
         for name in self._effect_cache:
             cache = self._effect_cache[name]
-            for spec_str, (effect_id, requirements) in cache.items():
+            for (spec_str, _), (effect_id, requirements) in cache.items():
                 self.gen.fact(fn.pkg_fact(name, fn.effect_id(effect_id)))
                 self.gen.fact(fn.pkg_fact(name, fn.effect_msg(spec_str)))
                 for predicate in requirements:
@@ -1477,71 +1484,70 @@ def variant_rules(self, pkg):
 
             self.gen.newline()
 
-    def condition(self, required_spec, imposed_spec=None, name=None, msg=None, node=False):
+    def condition(
+        self,
+        required_spec: spack.spec.Spec,
+        imposed_spec: Optional[spack.spec.Spec] = None,
+        name: Optional[str] = None,
+        msg: Optional[str] = None,
+        transform_required: Optional[TransformFunction] = None,
+        transform_imposed: Optional[TransformFunction] = remove_node,
+    ):
         """Generate facts for a dependency or virtual provider condition.
 
-        Constraints can be passed as a Spec, or as a list of AspFunction objects constructed
-        manually using self.spec_clauses(). The latter format is supported for cases where
-        individual constraints must be added or removed.
-
         Arguments:
-            required_spec (spack.spec.Spec or List[AspFunction]): the constraints that triggers
-                this condition
-            imposed_spec (spack.spec.Spec, List[AspFunction], or None): the constraints that are
-                imposed when this condition is triggered
-            name (str or None): name for `required_spec` (required if
-                required_spec is anonymous, ignored if not)
-            msg (str or None): description of the condition
-            node (bool): if False does not emit "node" or "virtual_node" requirements
-                from the imposed spec
+            required_spec: the constraints that triggers this condition
+            imposed_spec: the constraints that are imposed when this condition is triggered
+            name: name for `required_spec` (required if required_spec is anonymous, ignored if not)
+            msg: description of the condition
+            transform_required: transformation applied to facts from the required spec. Defaults
+                to leave facts as they are.
+            transform_imposed: transformation applied to facts from the imposed spec. Defaults
+                to removing "node" and "virtual_node" facts.
         Returns:
             int: id of the condition created by this function
         """
-        # Defensive copy of string or spec
         named_cond = required_spec.copy()
-        if isinstance(named_cond, spack.spec.Spec):
-            named_cond.name = named_cond.name or name
-            name = named_cond.name
-        assert name, "must provide name for anonymous conditions!"
+        named_cond.name = named_cond.name or name
+        assert named_cond.name, "must provide name for anonymous conditions!"
+
         # Check if we can emit the requirements before updating the condition ID counter.
         # In this way, if a condition can't be emitted but the exception is handled in the caller,
         # we won't emit partial facts.
 
         condition_id = next(self._condition_id_counter)
-        self.gen.fact(fn.pkg_fact(name, fn.condition(condition_id)))
+        self.gen.fact(fn.pkg_fact(named_cond.name, fn.condition(condition_id)))
         self.gen.fact(fn.condition_reason(condition_id, msg))
 
-        cache = self._trigger_cache[name]
+        cache = self._trigger_cache[named_cond.name]
 
-        named_cond_key = str(named_cond)
+        named_cond_key = (str(named_cond), transform_required)
         if named_cond_key not in cache:
             trigger_id = next(self._trigger_id_counter)
-            if isinstance(named_cond, spack.spec.Spec):
-                requirements = self.spec_clauses(named_cond, body=True, required_from=name)
-            else:
-                requirements = named_cond
+            requirements = self.spec_clauses(named_cond, body=True, required_from=name)
+
+            if transform_required:
+                requirements = transform_required(named_cond, requirements)
+
             cache[named_cond_key] = (trigger_id, requirements)
         trigger_id, requirements = cache[named_cond_key]
-        self.gen.fact(fn.pkg_fact(name, fn.condition_trigger(condition_id, trigger_id)))
+        self.gen.fact(fn.pkg_fact(named_cond.name, fn.condition_trigger(condition_id, trigger_id)))
 
         if not imposed_spec:
             return condition_id
 
-        cache = self._effect_cache[name]
-        imposed_spec_key = str(imposed_spec)
+        cache = self._effect_cache[named_cond.name]
+        imposed_spec_key = (str(imposed_spec), transform_imposed)
         if imposed_spec_key not in cache:
             effect_id = next(self._effect_id_counter)
-            if isinstance(imposed_spec, spack.spec.Spec):
-                requirements = self.spec_clauses(imposed_spec, body=False, required_from=name)
-            else:
-                requirements = imposed_spec
-            if not node:
-                requirements = list(
-                    filter(lambda x: x.args[0] not in ("node", "virtual_node"), requirements)
-                )
+            requirements = self.spec_clauses(imposed_spec, body=False, required_from=name)
+
+            if transform_imposed:
+                requirements = transform_imposed(imposed_spec, requirements)
+
             cache[imposed_spec_key] = (effect_id, requirements)
         effect_id, requirements = cache[imposed_spec_key]
-        self.gen.fact(fn.pkg_fact(name, fn.condition_effect(condition_id, effect_id)))
+        self.gen.fact(fn.pkg_fact(named_cond.name, fn.condition_effect(condition_id, effect_id)))
         return condition_id
 
     def impose(self, condition_id, imposed_spec, node=True, name=None, body=False):
@@ -1587,24 +1593,32 @@ def package_dependencies_rules(self, pkg):
                 if not depflag:
                     continue
 
-                msg = "%s depends on %s" % (pkg.name, dep.spec)
+                msg = f"{pkg.name} depends on {dep.spec}"
                 if cond != spack.spec.Spec():
-                    msg += " when %s" % cond
+                    msg += f" when {cond}"
                 else:
                     pass
 
-                required_spec = cond.copy()
-                required_spec.name = cond.name or pkg.name
-                required = self.spec_clauses(required_spec, body=True)
-                required += [fn.attr("track_dependencies", pkg.name)]
-                imposed = self.spec_clauses(dep.spec)
-                imposed += [
-                    fn.attr("dependency_holds", pkg.name, dep.spec.name, dt.flag_to_string(t))
-                    for t in dt.ALL_FLAGS
-                    if t & depflag
-                ]
+                def track_dependencies(input_spec, requirements):
+                    return requirements + [fn.attr("track_dependencies", input_spec.name)]
+
+                def dependency_holds(input_spec, requirements):
+                    return remove_node(input_spec, requirements) + [
+                        fn.attr(
+                            "dependency_holds", pkg.name, input_spec.name, dt.flag_to_string(t)
+                        )
+                        for t in dt.ALL_FLAGS
+                        if t & depflag
+                    ]
 
-                self.condition(required, imposed, pkg.name, msg)
+                self.condition(
+                    cond,
+                    dep.spec,
+                    name=pkg.name,
+                    msg=msg,
+                    transform_required=track_dependencies,
+                    transform_imposed=dependency_holds,
+                )
 
                 self.gen.newline()
 
@@ -1699,11 +1713,16 @@ def emit_facts_from_requirement_rules(self, rules: List[RequirementRule]):
                     when_spec = spack.spec.Spec(pkg_name)
 
                 try:
+                    # With virtual we want to emit "node" and "virtual_node" in imposed specs
+                    transform: Optional[TransformFunction] = remove_node
+                    if virtual:
+                        transform = None
+
                     member_id = self.condition(
                         required_spec=when_spec,
                         imposed_spec=spec,
                         name=pkg_name,
-                        node=virtual,
+                        transform_imposed=transform,
                         msg=f"{spec_str} is a requirement for package {pkg_name}",
                     )
                 except Exception as e:
@@ -1767,8 +1786,16 @@ def external_packages(self):
             # Declare external conditions with a local index into packages.yaml
             for local_idx, spec in enumerate(external_specs):
                 msg = "%s available as external when satisfying %s" % (spec.name, spec)
-                external_imposition = [fn.attr("external_conditions_hold", spec.name, local_idx)]
-                self.condition(spec, external_imposition, msg=msg)
+
+                def external_imposition(input_spec, _):
+                    return [fn.attr("external_conditions_hold", input_spec.name, local_idx)]
+
+                self.condition(
+                    spec,
+                    spack.spec.Spec(spec.name),
+                    msg=msg,
+                    transform_imposed=external_imposition,
+                )
                 self.possible_versions[spec.name].add(spec.version)
                 self.gen.newline()
 
@@ -2621,7 +2648,7 @@ def literal_specs(self, specs):
             self.gen.fact(fn.literal(trigger_id))
 
             # Effect imposes the spec
-            imposed_spec_key = str(spec)
+            imposed_spec_key = str(spec), None
             cache = self._effect_cache[spec.name]
             msg = (
                 "literal specs have different requirements. clear cache before computing literals"

```

</details>

If it's fine with you I can push that here, let me know.

alalazo committed:
Inject transformation functions into "condition"

By injecting transformation functions, we keep
required the input types the same as before.

The "node" argument is generalized to be a generic
transformation function that can manipulate the
requirements to be emitted

alalazo commented on lib/spack/spack/solver/error_messages.lp:
```diff
@@ -0,0 +1,216 @@
+%
+% SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+%=============================================================================
```
In this file we need to add a `#defined ` for at least these two facts:
```
/home/culpo/PycharmProjects/spack/lib/spack/spack/solver/error_messages.lp:92:10-39: info: atom does not occur in any rule head:
  provider(#P0,node(#X1,#X2))

/home/culpo/PycharmProjects/spack/lib/spack/spack/solver/error_messages.lp:121:10-51: info: atom does not occur in any rule head:
  external_version(node(#X0,#X1),#P2,#P3)
```

alalazo commented on lib/spack/spack/solver/asp.py:
```diff
@@ -685,6 +685,7 @@ def extract_args(model, predicate_name):
 class ErrorHandler:
     def __init__(self, model):
         self.model = model
+        self.full_model = None
```
(minor) We can get rid of `self.errors` here. Since the attribute is only used in `raise_if_errors`, we can have a local variable in that method.

alalazo commented on lib/spack/spack/solver/asp.py:
```diff
@@ -693,6 +694,24 @@ def multiple_values_error(self, attribute, pkg):
     def no_value_error(self, attribute, pkg):
         return f'Cannot select a single "{attribute}" for package "{pkg}"'
 
+    # TODO GBB: make these only track within the same condition set
```
Is this for later of for this PR?

alalazo commented on lib/spack/spack/solver/asp.py:
```diff
@@ -693,6 +694,24 @@ def multiple_values_error(self, attribute, pkg):
         return f'Cannot select a single "{attribute}" for package "{pkg}"'
 
+    # TODO GBB: make these only track within the same condition set
+    def _get_cause_tree(self, cause, conditions, condition_causes, literals, indent="        "):
```
Do we have ways to prevent cycles to cause infinite recursions here? Thinking of cases where:
1. `a` imposes `b`
2. `b` imposes `a`

And the cycle is triggered in some way.

alalazo replied:
Also, can we add docstrings and type hints for these functions?

alalazo commented on lib/spack/spack/solver/error_messages.lp:
```diff
@@ -0,0 +1,216 @@
+
+% More specific error message if the version cannot satisfy some constraint
+% Otherwise covered by `no_version_error` and `versions_conflict_error`.
+error(1, "Cannot satisfy '{0}@{1}'", Package, Constraint, startcauses, ConstraintCause)
```
We should write somewhere clear that `startcauses` is a placeholder value for error reconstruction in ASP. It's otherwise hard to follow what's going on the first time.

alalazo reviewed (commented):
Ok, finished first complete re-review

becker33 committed:
ensure condition associations don't jump condition sets

becker33 committed:
define provider and external_version in error_messages.lp

becker33 committed:
refactor error handling for less global state

becker33 committed:
document startcauses in error_messages.lp

becker33 committed:
docstrings and type hints for get_cause_tree methods

becker33 committed:
cycle avoidance in get_cause_tree methods

becker33 committed:
ensure condition causes don't bleed across condition sets

becker33 committed:
bugfix in requirement for node_flag_source

becker33 committed:
fixup 7659db435d

becker33 committed:
Merge branch 'develop' into features/error-message-chaining-3

becker33 committed:
Add tests for causal error messages

becker33 committed:
style

alalazo commented:
@spackbot run pipeline

alalazo commented:
There's a pipeline that seems to consistently fail at concretization in this PR: 
- https://gitlab.spack.io/spack/spack/-/jobs/8971362

becker33 committed:
test higher memory request for CI

alalazo committed:
environment: solve one spec per child process

Looking at the memory profiles of concurrent solves
for environment with unify:false, it seems memory
is only ramping up.

This exchange in the potassco mailing list:
 https://sourceforge.net/p/potassco/mailman/potassco-users/thread/b55b5b8c2e8945409abb3fa3c935c27e%40lohn.at/#msg36517698

Seems to suggest that clingo doesn't release memory
until end of the application.

Since when unify:false we distribute work to processes,
here we give a maxtaskperchild=1, so we clean memory
after each solve.

alalazo commented:
Let's see if https://github.com/spack/spack/pull/40876 helps lowering the memory requests during concretization.

becker33 committed:
Revert "test higher memory request for CI"

This reverts commit 99d89f3897329b5c1cf1db5cb9918f5df9450dca.

alalazo commented:
So, we want back to a quite noticeable general slowdown (probably due to rebasing on cherry-pick vdeps):

[radiuss-pr.csv](https://github.com/spack/spack/files/13256162/radiuss-pr.csv)
[radiuss-develop.csv](https://github.com/spack/spack/files/13256163/radiuss-develop.csv)
[radiuss.txt](https://github.com/spack/spack/files/13256164/radiuss.txt)

![radiuss](https://github.com/spack/spack/assets/4199709/dc8076ab-7f57-4224-b44b-43e76d7e31c9)

Need to investigate what caused the issue, which is mostly in grounding.

tgamblin commented:
So neither of these caused a slowdown on its own, but together they did?  Curious what they have in common that could cause the slowdown at ground time.

alalazo commented on lib/spack/spack/solver/concretize.lp:
```diff
@@ -115,29 +115,51 @@ multiple_nodes_attribute("depends_on").
+  trigger_and_effect(Root, TriggerID, EffectID),
+  imposed_constraint(EffectID, "provider_set", _, Mentioned).
+
+condition_set(node(min_dupe_id, Root), node(min_dupe_id, Mentioned), link_run) :-
+  mentioned_in_literal(Root, Mentioned).
```
I think this is wrong. For instance with a spec like:
```
hdf5 ^cmake@3.26
```
`cmake` is not a link-run dependency in the condition set of `hdf5`. Also, I should have deleted `link_run` since we are using only `direct_link_run` and `direct_build`.

alalazo replied:
See https://github.com/spack/spack/pull/40173/commits/67dd6cebd5557a17647928c0e0122c7dce0049db (I have to think hard if there are edge cases where computing `/2` facts directly for roots may be wrong).

alalazo commented on lib/spack/spack/solver/concretize.lp:
```diff
@@ -373,6 +397,14 @@ attr("node_flag_source", node(X, A1), A2, node(Y, A3))
      imposed_constraint(ID, "node_flag_source", A1, A2, A3),
      condition_set(node(Y, A3), node(X, A1)).
 
+attr("provider_set", node(PDupID, Provider), node(VDupID, Virtual))
+  :- impose(ID, RequestingNode),
+     imposed_constraint(ID, "provider_set", Provider, Virtual),
+     condition_set(RequestingNode, node(PDupID, Provider)),
+     condition_set(RequestingNode, node(VDupID, Virtual)).
```
I suspect this rule is what increased the solution time the most.

alalazo replied:
Restricted the scope of this rule in https://github.com/spack/spack/pull/40173/commits/b10d54b5e4ef96a4650db9a559ad7721f9ac93a8

alalazo committed:
Construct condition_set/2 directly from literals

alalazo committed:
Construct the condition_set only when the root is effectively solved

alalazo committed:
Simplify computation of `mentioned_in_literal` facts

alalazo committed:
Reduce scope of "provider_set" fact

alalazo commented:
I have been able to improve the performance regression, but it's still not perfect. On:

* **Spack:** 0.21.0.dev0 (b10d54b5e4ef96a4650db9a559ad7721f9ac93a8)
* **Python:** 3.11.5
* **Platform:** linux-ubuntu20.04-icelake
* **Concretizer:** clingo

I got these result:

[radiuss-pr.csv](https://github.com/spack/spack/files/13265579/radiuss-pr.csv)
[radiuss-develop.csv](https://github.com/spack/spack/files/13265580/radiuss-develop.csv)
[radiuss.txt](https://github.com/spack/spack/files/13265581/radiuss.txt)

![radiuss](https://github.com/spack/spack/assets/4199709/967fd29c-a518-49a6-aaad-cd40102bccb8)

So, we got 10-20% generalized slowdown - I'll check what I can improve next.
